styx-scheduler-service/src/main/java/com/spotify/styx/workflow/WorkflowInitializer.java

```diff
@@ -46,32 +49,43 @@ public void inspectChange(Workflow workflow) {
     final Optional<Workflow> previous;
     try {
       previous = storage.workflow(workflow.id());
-      storage.storeWorkflow(workflow);
     } catch (IOException e) {
+      LOG.warn("failed to read workflow {} from storage", workflow.id(), e);
       throw new RuntimeException(e);
     }
 
+    Optional<TriggerInstantSpec> nextSpec = Optional.empty();
+
     // either the workflow is completely new, or the schedule/offset has changed
     final Schedule newSchedule = workflow.configuration().schedule();
     final Optional<String> newOffset = workflow.configuration().offset();
     if (!previous.isPresent()
         || !previous.get().configuration().schedule().equals(newSchedule)
         || !previous.get().configuration().offset().equals(newOffset)) {
-      initializeNaturalTrigger(workflow);
+      try {
+        nextSpec = Optional.of(initializeNaturalTrigger(workflow));
+      } catch (Exception e) {
+        LOG.info("could not compute next natural trigger for workflow {}", workflow, e);
+        throw new WorkflowInitializationException(e);
+      }
+    }
+
+    try {
+      storage.storeWorkflow(workflow);
+      if (nextSpec.isPresent()) {
+        storage.updateNextNaturalTrigger(workflow.id(), nextSpec.get());
```

Comment:
I wonder why there's no "racy" comment here. If we storeWorkflow but fail to updateNextNaturalTrigger, then I guess the workflow stays in Styx as a zombie.

Anyway seems a minor concern, and not introduced in this PR.

Reply:
It's not strictly a zombie. If the workflow was there, the worst case is next natural trigger not updated; if it was not there, we will have an empty next natural trigger. Either case, the client will receive 500.